### ISSUE TITLE
How do create a language server with Surelog/UHDM

### ISSUE THREAD

**Issue by alaindargelas** [chipsalliance/Surelog#1712]
From @hs-apotell:

A little context here would have helped. My bad.

This issue came up because one of team member is attempting to get a language server working and he has a very basic functional prototype to showcase. It was just by accident that he was using a rather large design for test case and the response time was very slow. So, he decided to fallback to load just the cache file, which worked well (for the minimum feature set he was focusing on) until he pulled and refreshed his local repository.

Long story short, if we are working with rather large design and we know for sure that only a specific file has been modified and need to be refreshed in the cache, can this be done safely. This thread is the result of that discussion.

So, coming back to the need for fast response time that is required for a language server like utility to work, what can be done to avoid/prevent having to re-parse the entire design over and over again for every keystroke?

**Comment by alaindargelas** [comment]
Is this still relevant or you have moved on some other solution?

**Comment by hs-apotell** [comment]
Yes. There is work in progress on this just not a very high priority at the moment for us. Unfortunately, the core developer working on this had to move on to something more pressing. The core issue is however still relevant.

**Comment by hzeller** [comment]
For other implementations: I've implemented a SystemVerilog language server as one of the tools in Verible, which at this point does syntax checking, linting and formatting https://github.com/chipsalliance/verible/tree/master/verilog/tools/ls

**Comment by hs-apotell** [comment]
> BTW, are you going to contribute back the Linter or the LS? Have you noted the UhdmLinter and SynthSubset optional classes in UHDM?

Yes & No. Yes it will be open source once I have it working with demonstrable features. However, It may be part of Surelog. The reason being that I am employing two different parsers - Surelog & Verible. Surelog for static parsing i.e. all the files that aren't currently changing and Verible for dynamic parsing i.e. the ones that are actively being edited. I think this approach gets us best of both worlds. All the semantic information that Surelog can provide and the speed that is needed for dynamically changing code.

**Comment by alaindargelas** [comment]
Excellent!
